server/src/server.jl

```diff
@@ -0,0 +1,164 @@
+    "get_current_time" => GetCurrentTime,
+    "update_until" => UpdateUntil,
+    "update" => Update,
+    "get_input_var_names" => GetInputVarNames,
+    "get_output_var_names" => GetOutputVarNames,
+    "get_var_itemsize" => GetVarItemSize,
+    "get_var_type" => GetVarType,
+    "get_var_units" => GetVarUnits,
+    "get_var_location" => GetVarLocation,
+    "get_var_nbytes" => GetVarNbytes,
+    "get_value" => GetValue,
+    "get_value_ptr" => GetValuePtr,
+    "get_value_at_indices" => GetValueAtIndices,
+    "set_value" => SetValue,
+    "set_value_at_indices" => SetValueAtIndices,
+    "get_grid_type" => GetGridType,
+    "get_var_grid" => GetVarGrid,
+    "get_grid_rank" => GetGridRank,
+    "get_grid_size" => GetGridSize,
+    "get_grid_x" => GetGridX,
+    "get_grid_y" => GetGridY,
+    "get_grid_node_count" => GetGridNodeCount,
+    "get_grid_edge_count" => GetGridEdgeCount,
+    "get_grid_edge_nodes" => GetGridEdgeNodes,
+    "finalize" => Finalize,
+    "load_state" => LoadState,
+    "save_state" => SaveState,
+)
+
+mutable struct ModelHandler
+    model::Union{Wflow.Model,Nothing}
+end
+
+"Shutdown ZMQ server"
+function shutdown(s::Socket, ctx::Context)
+    @info "Shutting down Wflow ZMQ server on request..."
+    ZMQ.close(s)
+    ZMQ.close(ctx)
+end
+
+"Error response ZMQ server"
+function response(err::AbstractString, s::Socket)
+    @info "Send error response"
+    resp = Dict{String,String}("status" => "ERROR", "error" => err)
+    ZMQ.send(s, JSON3.write(resp))
+end
+
+"Status response ZMQ server"
+function response(s::Socket)
+    @info "Send status response"
+    resp = Dict{String,String}("status" => "OK")
+    ZMQ.send(s, JSON3.write(resp))
+end
+
+"Validate JSON request against mapped Struct"
+function valid_request(json)
+    for f in fieldnames(map_structs[json.fn])
+        if f ∉ keys(json)
+            return f
+            break
+        end
+    end
+end
+
+"""
+    wflow_bmi(s::Socket, handler::ModelHandler, f)
+
+Run a Wflow function through `wflow.bmi(f, handler.model)` and update Wflow Model `handler`
+if required, depending on return type of `wflow.bmi(f, handler.model)`.
+"""
+function wflow_bmi(s::Socket, handler::ModelHandler, f)
+    try
+        ret = wflow_bmi(f, handler.model)
+        if typeof(ret) <: Wflow.Model # update of Wflow model
+            handler.model = ret
+            response(s)
+        elseif isnothing(ret) # for SetValue and SetValueAtIndices
+            response(s)
+        else
+            @info "Send response including output from Wflow function `$(f.fn)`"
+            ZMQ.send(s, JSON3.write(ret))
+        end
+    catch e
+        @error "Wflow function `$(f.fn)` failed" exception = (e, catch_backtrace())
+        err = string(
+            "Wflow function `$(f.fn)` failed\n exception =\n ",
+            sprint(showerror, e, catch_backtrace()),
+        )
+        response(err, s)
+    end
+end
+
+"Start Wflow ZMQ Server"
+function start()
```

Comment:
```suggestion
function main(ARGS::Vector{String})
```

To future-proof this I'd go with the standardized entry point `main(ARGS)`.
Potentially this will also apply to modules.

https://github.com/JuliaLang/julia/blob/master/NEWS.md#command-line-option-changes

Or have both, and have `main(ARGS::Vector{String})` call `start(port::Int)`, the former being the entry point and the latter the Julia API.
See also https://github.com/Deltares/Ribasim/blob/main/core/src/main.jl

Reply:
Thanks for this suggestion and the web links, I went for both, see commit https://github.com/Deltares/Wflow.jl/pull/356/commits/0104c187adb5a656906aeae6318fb44dad68b6f6.